Feature/300 create advanced tutorials 2 (#14)
G1ANT-Robot/G1ANT.Tutorials: Automation Examples/020.md

# Process

Open https://www.vat19.com/, go to fun & strange stuff and download first 20 products' information (name, price, image) and save it into excel file. Template is available [here](Products%20Template.xlsx).

# Steps to automate

#### 1. Specify a path to your folder, open website and maximise the browser

> 

Author: ```suggestion

Remember to select the selenium addon in the Addons panel (View/Addons) to be able to use selenium commands.
```

```G1ANT
♥basePath = C:\Users\dell\Desktop\Automation Examples

> selenium.open chrome url https://www.vat19.com/

Reviewer: usage of selenium addon: not mentioned to include addon to robot.

Author: Grrrrr 😛

window ✱Unique✱ style maximize
```

> Use window command and part of the window title as an argument to bring it to the front and maximise.

Reviewer: `and part of the window title as an argument` i can't see references to the window title in code, am I missing something?
![image](https://user-images.githubusercontent.com/7533717/64418522-74d2ef80-d09b-11e9-8295-fafaf4524a50.png)

Author: ![chrome_2019-09-06_12-07-16](https://user-images.githubusercontent.com/39299658/64420164-ed877b00-d09e-11e9-8980-72988ab3d38d.png)

Reviewer: hmm this feels strange, really. imho tutorial should be 100% repeatable

Author: Because you weren't at the main webpage, you opened Fun & Strange Stuff

#### 2. Go to fun & strange stuff, create lists which will contain product names and their prices

> After running the above script and seeing the webpage, you can inspect the button element and in the developer tools, right click it and choose "Copy Xpath" that you can insert in the search argument for selenium.click command.

Reviewer: I don't see any button element.. :( Elaborate on this; it's hard to find out what to do.
![image](https://user-images.githubusercontent.com/7533717/64418971-3be74a80-d09c-11e9-90bd-155a9839b6aa.png)

Author: In Automation Anywhere example it wasn't described. Users are told to open "Fun & Strange Stuff" webpage so it's obvious that you need to click this button while being on the main webpage.

Reviewer: i'm just saying that it is not a *button*, it's just navigation menu item

Author: ```suggestion
After running the above script and seeing the webpage, you can inspect the "Fun & Strange Stuff" element and in the developer tools, right click it and choose "Copy Xpath" that you can insert in the search argument for selenium.click command.
```

```G1ANT
selenium.click search ‴//*[@id="side-nav-content"]/div[1]/ul/li[3]/a‴ by xpath
list.create type text result ♥products
list.create type text result ♥prices
```

#### 3. Wait for elements to be loaded, get link to the first image, get name of the product and price value and store them in the lists

> Right click the image element and see that the "src" attribute stores a link to it. The below JavaScript code returns a link to the first image. You can run it in the console to check it. 

Reviewer: ```suggestion
Right click the image element and see that the "src" attribute stores a link to it. The below JavaScript code returns a link to the first image. You can run it in the browser's console to check it. 
```

```js
document.getElementsByClassName("lazyautosizes lazyloaded")[0].getAttribute("src")
```

So let's wait for this element using the below G1ANT code.

```G1ANT
selenium.waitforvalue ‴return document.getElementsByClassName("lazyautosizes lazyloaded").length > 0‴ expectedvalue true
```

After right clicking the product name and price, you may see that geting those values is as easy as below. 

```G1ANT
selenium.runscript ‴return document.getElementsByClassName("productSketch-name-link")[0].innerText‴ result ♥productName
list.add ♥products toadd ♥productName
selenium.runscript ‴return document.getElementsByClassName("regularPrice")[0].innerText‴ result ♥price
list.add ♥prices toadd ♥price
```

> #### 4. Donwload the first image

Reviewer: ```suggestion
#### 4. Download the first image
```

In order to do this, add "https:" prefix to the "src" and use file.download command.

```G1ANT
♥imageSrc = ⊂"https:" + ♥imageSrc⊃
file.download ♥imageSrc filename ♥basePath\1.png
```

#### 5. Put all above steps in a loop

This is what you should get:

```G1ANT
for ♥i from 0 to 19
    selenium.runscript ‴return document.getElementsByClassName("lazyautosizes lazyloaded")[♥i].getAttribute("src")‴ result ♥imageSrc
    ♥imageSrc = ⊂"https:" + ♥imageSrc⊃
    file.download ♥imageSrc filename ♥basePath\⊂♥i+1⊃.png
    selenium.runscript ‴return document.getElementsByClassName("productSketch-name-link")[♥i].innerText‴ result ♥productName
    list.add ♥products toadd ♥productName
    selenium.runscript ‴return document.getElementsByClassName("regularPrice")[♥i].innerText‴ result ♥price
    list.add ♥prices toadd ♥price
end for
```

#### 6. After running this script, you may notice that it fails after some time

> It does so because webpage isn't fully loaded and some of the elements are not yet present so the script can't continue. To load more elements, you should scroll. Insert below line at the end of the loop to prevent this.

Reviewer: "to prevent this"? 
i'd rather say "to achieve this"

Author: yeah hahahaha "mind shortcut" XD

Author: ```suggestion
It does so because webpage isn't fully loaded and some of the elements are not yet present so the script can't continue. To load more elements, you should scroll. Insert below line at the end of the loop to achieve this.
```

```G1ANT

>     selenium.runscript ‴window.scrollBy(0, 200)‴

Reviewer: why 200?

Author: just an example

```

#### 7. Close the browser and open excel template

```G1ANT
selenium.close

excel.open ‴♥basePath\Images Template.xlsx‴
```

#### 8. Create a loop, insert images, products and prices

After pressing alt in the excel application, some of the letters appear which indicates what shortcut to use if you want to open some specific option from the menu. In this case, we need to press "n" and "p" to insert an image. Using keyboard to insert long texts letter by letter is sometimes not the fastest option so let's use clipboard for that and Ctrl+V shortuct to paste it in the "Save As" window. Then, set values for product name and price. Lastly, go the first column, next row by using Home and Down keys.

```G1ANT
for ♥i from 1 to 20

>     keyboard ⋘alt⋙np

Reviewer: ⋘alt⋙np
shouldn't it be more like ⋘alt+n+p⋙?

Author: No. To access something from the menu, you need to press alt so that letters are underlined and focused, then "n" to open "Insert" tab, then "P". It is crucial because sometimes to access some option, you have two letters next to each other (which means that they have to be pressed at the same time). See next example where in order to open "Format" tab, you need to press "J+P"
